Enhance FilterBar with DateRangeFilter for compact UI

RoyLee1224 opened the pull request:
## Why

`logical date` and `run after` from/to filters are taking too much space in both filter bar & dropdown

Switching to a range filter saves space, aligns better with the design of other components, and provides a more convenient user experience.


## To-Do

- [x] Added `DateRangeFilter` & `DateRangeCalendar` components.
- [x] Add utility functions: `getDefaultFilterValue` and `isValidFilterValue`.
- [x] Define date format (YYYY/MM/DD) as constant
- [x] Replace hardcoded colors
- [x] Support dark mode
- [x] Let users include time when selecting range
- [x] Add tests for Date Range
- [x] Implement range filter for `Events`, `AssetEvents`, `DagRuns`,
- [x] Implement input validation
- [x] Remove from/to filters
- [x] Test `datetime-local` issues for Safari and Firefox

## Note
**Datetime Input Validation**
  - Same Day: Start time must be ≤ end time (e.g., 10:00 → 20:00 ✅, 20:00 → 10:00 ❌)
  - Cross Day: Any time combination is valid (e.g., 2025/11/03 23:00 → 2025/11/04 08:00 ✅)  

## Screenshots
### Dropdown before / after
<img width="299" height="452" alt="dropdown_before" src="https://github.com/user-attachments/assets/7cb77815-9c53-40a9-9dd5-45bba4d2ea94" />
<img width="299" height="452" alt="dropdown_after" src="https://github.com/user-attachments/assets/d2bb6cc3-57ba-4fb3-8415-156f477f3f7e" />

### Date picker before / after
<img width="900" height="602" alt="CleanShot 2025-11-07 at 15 59 11" src="https://github.com/user-attachments/assets/dc40cd76-a39e-4693-a8b8-20ec13bc4be4" />
<img width="900" height="602" alt="CleanShot 2025-11-07 at 15 58 21" src="https://github.com/user-attachments/assets/dd1d0d77-d743-4227-b252-0b34de71a2ee" />


### Test for DateRangeFilter (DateRangeFilter.test.tsx)
<img width="2432" height="738" alt="CleanShot 2025-11-06 at 19 41 10@2x" src="https://github.com/user-attachments/assets/be0cf749-9129-4773-8d44-8a2f63874cab" />



<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

RoyLee1224 committed:
feat(ui): Add date range filter

pierrejeambrun commented on airflow-core/src/airflow/ui/public/i18n/locales/en/common.json:
```diff
@@ -70,6 +70,7 @@
   "dagRun_other": "Dag Runs",
   "dagRunId": "Dag Run ID",
   "dagWarnings": "Dag warnings/errors",
+  "date": "Date",
```
I would keep `Logical Date` especially to because it's displayed next to `Run After Date`, just to be more explicit.

RoyLee1224 replied:
Sure, I'm on it

pierrejeambrun reviewed (commented):
That's a good idea. We probably want to also address other places where we have date range filtering. For instance `Events` `DagRuns`.

Are we still able to specify one end of the range without specifying the other? (For instance specify the start date but not the end date).

pierrejeambrun commented on airflow-core/src/airflow/ui/src/components/FilterBar/filters/TextSearchFilter.tsx:
```diff
@@ -56,7 +57,7 @@ export const TextSearchFilter = ({ filter, onChange, onRemove }: FilterPluginPro
         onChange={handleInputChange}
         placeholder={filter.config.placeholder}
-        value={String(filter.value ?? "")}
+        value={typeof filter.value === "string" ? filter.value : ""}
```
Why did we change the way we handle strings, here and in other files too? That doesn't seem needed for the PR.

RoyLee1224 replied:
@pierrejeambrun I think the string handling changes were necessary because `FilterValue` now includes `DateRangeValue` objects. I'm getting ESLint errors without these changes.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/constants/filterConfigs.tsx:
```diff
@@ -85,6 +86,11 @@ export const useFilterConfigs = () => {
       label: translate("common:filters.logicalDateTo"),
       type: FilterTypes.DATE,
     },
+    [SearchParamsKeys.LOGICAL_DATE_RANGE]: {
+      icon: <MdDateRange />,
+      label: translate("common:date"),
+      type: FilterTypes.DATERANGE,
+    },
```
You can remove 'non range' filters, as I think range will replace them?

RoyLee1224 replied:
Thanks for bringing this up, I was thinking the same thing.

I'll remove the date filters and implement date range filters for those.

RoyLee1224 replied:
Done! 👍

bbovenzi commented on airflow-core/src/airflow/ui/src/hooks/useCalendarGrid.ts:
```diff
@@ -0,0 +1,38 @@
+import type { Dayjs } from "dayjs";
+import { useMemo } from "react";
+
+export const useCalendarGrid = (currentMonth: Dayjs) =>
```
Nit: Since grid view is an important view, I think `useCalendarSelect` may be better

RoyLee1224 replied:
Agree, working on it

bbovenzi commented on airflow-core/src/airflow/ui/src/components/FilterBar/filters/DateRangeCalendar.tsx:
```diff
@@ -0,0 +1,182 @@
+          return "blue.300";
+        }
+        if (state.isInRange) {
+          return "blue.50";
```
These should probably be `brand.*` colors. Could you provider a dark mode screenshot too?

RoyLee1224 replied:
Sure, I'm working on the dark mode. I'll provide screenshots once they are ready

bbovenzi commented on airflow-core/src/airflow/ui/src/components/FilterBar/filters/DateRangeFilter.tsx:
```diff
@@ -0,0 +1,151 @@
+  const endDateValue = useMemo(
+    () => (isValidDateValue(value.endDate) ? dayjs(value.endDate) : undefined),
+    [value.endDate],
+  );
```
Do we need to memoize these?

RoyLee1224 replied:
Nice catch, on it

bbovenzi commented on airflow-core/src/airflow/ui/src/components/FilterBar/filters/DateRangeFilter.tsx:
```diff
@@ -0,0 +1,151 @@
+              <Text color={hasValue ? "inherit" : "gray.500"} fontSize="sm">
+                {hasValue ? formatDisplayValue() : translate("common:filters.selectDateRange")}
+              </Text>
+              <Box
```
This should be made into a <Button> instead

RoyLee1224 replied:
Umm I think the message got truncated. Could you help me understand what approach you'd recommend for?

bbovenzi commented on airflow-core/src/airflow/ui/src/components/FilterBar/filters/DateRangeFilter.tsx:
```diff
@@ -0,0 +1,151 @@
+          </Box>
+        </Popover.Trigger>
+        <Popover.Content p={3} w="300px">
+          {renderCalendar()}
```
Let's not call rendercalendar. Let's just put the jsx component here.

RoyLee1224 replied:
Sure

bbovenzi commented on airflow-core/src/airflow/ui/src/components/FilterBar/filters/DateRangeInputs.tsx:
```diff
@@ -0,0 +1,134 @@
+        </Text>
+        <Input
+          _focus={{ borderColor: "blue.500" }}
+          borderColor={editingState.selectionTarget === "end" ? "blue.500" : "gray.300"}
```
We have a lot of manual numbered colors across this PR. We really should double check which of these are necessary and which can we just update a colorPalette or use semantic tokens

bbovenzi commented on airflow-core/src/airflow/ui/src/hooks/useDateRangeFilter.ts:
```diff
@@ -0,0 +1,143 @@
+      inputs: { ...prev.inputs, [field]: inputValue },
+    }));
+
+    const parsedDate = dayjs(inputValue, "YYYY/MM/DD", true);
```
We repeat this format a lot. Maybe it should be a const?

RoyLee1224 replied:
Good idea! Working on it.

bbovenzi reviewed (commented):
I really like this. We just need to refine it a bit more.

Two other asks:

1. Could you rebase so that I can run it locally without needing to rebuild?

2. We should try to add some tests for Date Range

RoyLee1224 committed:
refactor: improve date range filter and calendar components

  - Remove unnecessary useMemo
  - Extract DATE_INPUT_FORMAT
  - replace renderCalendar function
  - Remove unused useMemo import

RoyLee1224 committed:
refactor: replace hardcoded colors

RoyLee1224 committed:
refactor: implement config-driven date range filter handling

RoyLee1224 committed:
ui: Add Date Time Range Inputs

RoyLee1224 committed:
test: add test for DateRangeFilter

RoyLee1224 committed:
ui: implement range filter for Assets, DagRuns, Events

RoyLee1224 committed:
feat: add input validation for DateRangeFilter

RoyLee1224 committed:
refactor: improve DateRangeFilter with responsive design and component extraction

  - Increase filter pill width from 360px to responsive (min 300px, max 600px)
  - Extract DateInput component to reduce code duplication
  - Fix JSX depth issues and lint violations

RoyLee1224 committed:
fix: update test for date range filter

RoyLee1224 committed:
feat: add CREATED_AT_RANGE date filter for HITL pages

RoyLee1224 committed:
refactor: remove individual date filters in favor of range filters

RoyLee1224 committed:
style: show red border on date inputs when range validation fails

RoyLee1224 committed:
fix: clear validation errors when correcting date range via calendar & modify dropdown positioning

RoyLee1224 committed:
refactor(ui): Centralize filter initialValues logic in useFiltersHandler

RoyLee1224 committed:
fix: remove eslint-disable max-lines to pass static check

bbovenzi reviewed (approved):
Looks good! Might need a few UX improvements as we use it more but let's get this merged.
